logstash-core/src/main/java/org/logstash/ext/JrubyMemoryReadClientExt.java

```diff
@@ -0,0 +1,179 @@
+                ((RubyNumeric) waitForMillis).getIntValue(), TimeUnit.MILLISECONDS);
+    }
+
+    @JRubyMethod(name = "close")
+    public void rubyClose(final ThreadContext context) {
+        // noop, for compatibility with acked queue read client
+    }
+
+    public boolean isEmpty() {
+        return queue.isEmpty();
+    }
+
+    @JRubyMethod(name = "empty?")
+    public IRubyObject rubyIsEmpty(final ThreadContext context) {
+        return context.runtime.newBoolean(isEmpty());
+    }
+
+    public void setBatchDimensions(int batchSize, int waitForMillis) {
+        this.batchSize = batchSize;
+        waitForNanos = TimeUnit.NANOSECONDS.convert(waitForMillis, TimeUnit.MILLISECONDS);
+    }
+
+    @JRubyMethod(name = "set_batch_dimensions")
+    public IRubyObject rubySetBatchDimensions(final ThreadContext context, IRubyObject batchSize,
+                                              IRubyObject waitForMillis) {
+        setBatchDimensions(((RubyNumeric) batchSize).getIntValue(),
+                ((RubyNumeric) waitForMillis).getIntValue());
+        return this;
+    }
+
+    @JRubyMethod(name = "set_events_metric", required = 1)
+    public IRubyObject setEventsMetric(final ThreadContext context, IRubyObject metric) {
+        eventMetricOut = getCounter(metric, OUT_KEY);
+        eventMetricFiltered = getCounter(metric, FILTERED_KEY);
+        eventMetricTime = getCounter(metric, DURATION_IN_MILLIS_KEY);
+        return this;
+    }
+
+    @JRubyMethod(name = "set_pipeline_metric", required = 1)
+    public IRubyObject setPipelineMetric(final ThreadContext context, IRubyObject metric) {
+        pipelineMetricOut = getCounter(metric, OUT_KEY);
+        pipelineMetricFiltered = getCounter(metric, FILTERED_KEY);
+        pipelineMetricTime = getCounter(metric, DURATION_IN_MILLIS_KEY);
+        return this;
+    }
+
+    @JRubyMethod(name = "inflight_batches")
+    public IRubyObject rubyGetInflightBatches(final ThreadContext context) {
+        return RubyHash.newHash(context.runtime, inflightBatches, RubyBasicObject.UNDEF);
+    }
+
+    // create a new, empty batch
+    @JRubyMethod(name = "new_batch")
+    public IRubyObject newBatch(final ThreadContext context) {
+        return JrubyMemoryReadBatchExt.create();
+    }
+
+    @JRubyMethod(name = "read_batch")
+    public IRubyObject readBatch(final ThreadContext context) throws InterruptedException {
+        JrubyMemoryReadBatchExt batch = JrubyMemoryReadBatchExt.create(
+                LsQueueUtils.drain(queue, batchSize, waitForNanos));
+        startMetrics(batch);
+        return batch;
+    }
+
+    @JRubyMethod(name = "start_metrics")
+    public IRubyObject rubyStartMetrics(final ThreadContext context, IRubyObject batch) {
+        startMetrics((JrubyMemoryReadBatchExt) batch);
+        return this;
+    }
+
+    private void startMetrics(JrubyMemoryReadBatchExt batch) {
+        inflightBatches.put(Thread.currentThread().getId(), batch);
```

Comment:
@danhermann 

small improvement if you have a sec :) Just cache:

```java
Thread.currentThread().getId()
```

to a variable :)